numba/core/serialize.py

```diff
@@ -110,3 +131,375 @@ def _rebuild_code(marshal_version, bytecode_magic, marshalled):
+
+
+def _unpickle__CustomPickled(serialized):
+    """standard unpickling for `_CustomPickled`.
+
+    Uses `NumbaPickler` to load.
+    """
+    ctor, states = loads(serialized)
+    return _CustomPickled(ctor, states)
+
+
+def _pickle__CustomPickled(cp):
+    """standard pickling for `_CustomPickled`.
+
+    Uses `NumbaPickler` to dump.
+    """
+    serialized = dumps((cp.ctor, cp.states))
+    return _unpickle__CustomPickled, (serialized,)
+
+
+# Register custom pickling for the standard pickler.
+copyreg.pickle(_CustomPickled, _pickle__CustomPickled)
+
+
+def custom_reduce(cls, states):
+    """For customizing object serialization in `__reduce__`.
+
+    Object states provided here is used as keyword arguments to `._rebuild()`
+    class method.
+
+    Parameters
+    ----------
+    states : dict
+        Dictionary of object states to be serialized.
+
+    Returns
+    -------
+    result : tuple
+        This tuple conforms to the return type requirement for `__reduce__`.
+    """
+    return custom_rebuild, (_CustomPickled(cls, states),)
+
+
+def custom_rebuild(custom_pickled):
+    """Customized object deserialization.
+
+    This function is referenced internally by `custom_reduce()`.
+    """
+    cls, states = custom_pickled.ctor, custom_pickled.states
+    return cls._rebuild(**states)
+
+
+def check(obj):
+    """Check if *obj* can be serialized.
+
+    Parameters
+    ----------
+    obj : object
+
+    Returns
+    --------
+    can_serialize : bool
+    """
+    with io.BytesIO() as fout:
+        pickler = NumbaPickler(fout)
+        try:
+            pickler.dump(obj)
+        except pickle.PicklingError:
+            return False
+        else:
+            return True
+
+
+class _TracedPicklingError(pickle.PicklingError):
+    """A custom pickling error used internally in NumbaPickler.
+    """
+    pass
+
+
+class SlowNumbaPickler(pickle._Pickler):
+    """Extends the pure-python Pickler to support the pickling need in Numba.
+
+    Note: this is used on Python < 3.8 unless `pickle5` is installed.
+
+    Note: This is good for debugging because the C-pickler hides the traceback
+    """
+
+    dispatch = pickle._Pickler.dispatch.copy()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.__trace = []
+        self.__memo = {}
+
+    def save(self, obj):
+        """
+        Overrides parent's `.save()` to provide better error messages.
+        """
+        self.__trace.append(f"{type(obj)}: {id(obj)}" )
+        try:
+            return super().save(obj)
+        except _TracedPicklingError:
+            raise
+        except Exception as e:
+            def perline(items):
+                return '\n'.join(f" [{depth}]: {it}"
+                                 for depth, it in enumerate(items))
+            m = (f"Failed to pickle because of\n  {type(e).__name__}: {e}"
+                 f"\ntracing... \n{perline(self.__trace)}")
+            raise _TracedPicklingError(m)
```

Comment:
I won't think it is necessary. The `.__context__` will be set pointing to `e`. The `from e` will just set `__cause__`. `e` will be showed neither way for this case. Also, I found that `e` is not at all useful. The key info is knowing a certain type has failed pickling (almost always because of pickling is unsupported for that type) and knowing the object membership relationship of that problematic type to the top-level type being pickled.

Reply:
ok, agree, let's leave as is.